plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,375 @@
+
+-- ---------------------------------------------------------------------
+
+data ApplyOneParams = AOP
+  { file      :: Uri
+  , start_pos :: Position
+  -- | There can be more than one hint suggested at the same position, so HintTitle is used to distinguish between them.
+  , hintTitle :: HintTitle
+  } deriving (Eq,Show,Generic,FromJSON,ToJSON)
+
+type HintTitle = T.Text
+
+data OneHint = OneHint
+  { oneHintPos :: Position
+  , oneHintTitle :: HintTitle
+  } deriving (Eq, Show)
+
+applyOneCmd :: CommandFunction ApplyOneParams
+applyOneCmd lf ide (AOP uri pos title) = do
+  let oneHint = OneHint pos title
+  let file = maybe (error $ show uri ++ " is not a file.") toNormalizedFilePath'
+                   (uriToFilePath' uri)
+  let progTitle = "Applying hint: " <> title
+  withIndefiniteProgress lf progTitle Cancellable $ do
+    logm $ "hlint:applyOneCmd:file=" ++ show file
+    res <- applyHint ide file (Just oneHint)
+    logm $ "hlint:applyOneCmd:res=" ++ show res
+    return $
+      case res of
+        Left err -> (Left (responseError (T.pack $ "hlint:applyOne: " ++ show err)), Nothing)
+        Right fs -> (Right Null, Just (WorkspaceApplyEdit, ApplyWorkspaceEditParams fs))
+
+applyHint :: IdeState -> NormalizedFilePath -> Maybe OneHint -> IO (Either String WorkspaceEdit)
+applyHint ide nfp mhint =
+  runExceptT $ do
+    ideas <- bimapExceptT showParseError id $ ExceptT $ liftIO $ runAction "applyHint" ide $ getIdeas nfp
+    let ideas' = maybe ideas (`filterIdeas` ideas) mhint
+    let commands = map (show &&& ideaRefactoring) ideas'
+    liftIO $ logm $ "applyHint:apply=" ++ show commands
+    -- set Nothing as "position" for "applyRefactorings" because
+    -- applyRefactorings expects the provided position to be _within_ the scope
+    -- of each refactoring it will apply.
+    -- But "Idea"s returned by HLint point to starting position of the expressions
+    -- that contain refactorings, so they are often outside the refactorings' boundaries.
+    -- Example:
+    -- Given an expression "hlintTest = reid $ (myid ())"
+    -- Hlint returns an idea at the position (1,13)
+    -- That contains "Redundant brackets" refactoring at position (1,20):
+    --
+    -- [("src/App/Test.hs:5:13: Warning: Redundant bracket\nFound:\n  reid $ (myid ())\nWhy not:\n  reid $ myid ()\n",[Replace {rtype = Expr, pos = SrcSpan {startLine = 5, startCol = 20, endLine = 5, endCol = 29}, subts = [("x",SrcSpan {startLine = 5, startCol = 21, endLine = 5, endCol = 28})], orig = "x"}])]
+    --
+    -- If we provide "applyRefactorings" with "Just (1,13)" then
+    -- the "Redundant bracket" hint will never be executed
+    -- because SrcSpan (1,20,??,??) doesn't contain position (1,13).
+    let fp = fromNormalizedFilePath nfp
+    (_, mbOldContent) <- liftIO $ runAction "hlint" ide $ getFileContents nfp
+    oldContent <- maybe (liftIO $ T.readFile fp) return mbOldContent
+    res <- liftIO $ withSystemTempFile (takeFileName fp) $ \temp h -> do
```

Comment:
It's a bit sad that we need to hit the file system and reparse. Did you investigate using `runRefactoring` directly?

Reply:
Mmm taking a quick look to `runRefactoring`, i see it needs a filepath, so we still need to save the file if we want to ensure it contains the edited and not saved content (like we do now)
Moreover it needs the hints in a file, so we had to hit the filesystem twice.

Maybe we could check if any not saved edit has been made to not save a new file but use the original one. But not sure if that we will save us much time.

The right solution could be [change apply-refact](https://github.com/haskell/haskell-language-server/pull/166#issuecomment-682192837) to expose functions taking directly the content and use it. But i would do it after merging this.

Reply:
Added a comment about